### PR TITLE
feat: snapshot data provider update

### PR DESCRIPTION
**Pull request by sampolgar**
Update according to https://github.com/sismo-core/sismo-hub/issues/1431

### CODE REVIEW

**Comment by sampolgar** [comment]
Ah... I forgot to update the Interface Schema, I'll do that tomorrow

**Comment by sampolgar** [comment]
FYI @MartinGbz this has string params, I'll update it

**Comment by sampolgar** [comment]
I updated the params and interface

File: group-generators/helpers/data-providers/snapshot/provider.ts [hunk 1]
```diff
@@ -216,4 +226,461 @@ export default class SnapshotProvider
...
+    return Object.keys(authors).length;
+  }
+
+  /**
+   * Retrieves space admins from Snapshot.
+   * @param {string} string - space query parameter
+   * @returns {Promise<FetchedData>} - A Promise that resolves to an object containing space admins.
+   */
+  public async querySpaceAdmins({
+    space,
```
Comment (MartinGbz) [inline group-generators/helpers/data-providers/snapshot/provider.ts:383]: Same here: add a default value

File: group-generators/helpers/data-providers/snapshot/provider.ts [hunk 2]
```diff
@@ -216,4 +226,461 @@ export default class SnapshotProvider
...
+    return Object.keys(admins).length;
+  }
+
+  /**
+   * Retrieves space voters from Snapshot.
+   * This takes a long time because it needs to query all voters in SnapShot
+   *
+   * @param {QuerySpaceVotersAboveXInput} input - Contains parameters for the function:
+   *    @param {string} space - Optional - A string representing the space parameter of the query.
+   *    @param {number} abovex - Optional - A number that defines the minimum number of votes for an author to be included in the result.
```
Comment (MartinGbz) [inline group-generators/helpers/data-providers/snapshot/provider.ts:421]: like this no? => A number that defines the minimum number of votes for a **voter** to be included in the result.
\+ there are some empty space/indentation

File: group-generators/helpers/data-providers/snapshot/provider.ts [hunk 3]
```diff
@@ -216,4 +226,461 @@ export default class SnapshotProvider
...
+    return Object.keys(voters).length;
+  }
+
+  /**
+   * Retrieves proposal authors with votes above a specified number from a specified space and proposal state.
+   * This takes a long time because it needs to query all proposals in SnapShot
+   *
+   * @param {QueryProposalAuthorsAboveXInput} input - Contains parameters for the function:
+   *    @param {string} space - Optional - A string representing the space parameter of the query.
+   *    @param {number} abovex - Optional - A number that defines the minimum number of votes for an author to be included in the result.
```
Comment (MartinGbz) [inline group-generators/helpers/data-providers/snapshot/provider.ts:513]: like this no? => A number that defines the minimum number of **proposals created by** an author to be included in the result.

**Review by MartinGbz — COMMENTED** [review]
Thanks for this great PR again @sampolgar 💜

I add a few comments, but globally: can you add a defaultValue for each function, so anyone can if they want force a value for all the account fetch. And if this value is not set by the user keep the one you had planned (number of votes, number of proposals... etc)

Also, you just forgot to add your count function to the Data Provider index file (group-generators/helpers/data-providers/index.ts)

File: group-generators/helpers/data-providers/snapshot/provider.ts [hunk 4]
```diff
@@ -39,13 +47,12 @@ export default class SnapshotProvider
   }
 
   public async querySpaceVoters(
-    input: QuerySpaceVotersInput,
-    defaultValue = 1
```
[the remark this answers is not in the document]
Comment (sampolgar) [reply group-generators/helpers/data-providers/snapshot/provider.ts:43]: Thanks for pointing that out - I initially re-used this function but then decided to duplicate it and not use it. I forgot to change it back - I've changed it back.

[the remark this answers is not in the document]
Comment (sampolgar) [reply group-generators/helpers/data-providers/snapshot/provider.ts:43]: I was initially reusing the function in my queryspacevotersabovex, then I decided to recreate it a little bit differently. I forgot to change it back, I've done that now

File: group-generators/helpers/data-providers/snapshot/provider.ts (hunk 1 above)
Comment (sampolgar) [reply group-generators/helpers/data-providers/snapshot/provider.ts:383]: Thanks, I added the default value

File: group-generators/helpers/data-providers/snapshot/provider.ts (hunk 2 above)
Comment (sampolgar) [reply group-generators/helpers/data-providers/snapshot/provider.ts:421]: Yes - this was explicit in the GitHub issue so I included it
https://github.com/sismo-core/sismo-hub/issues/1431 

```
{
"name": "number of proposals",
"argName": "TODO",
"type": "TODO",
"example": "3",
"description": "A minimum number of proposals"
},
```

File: group-generators/helpers/data-providers/snapshot/provider.ts (hunk 3 above)
Comment (sampolgar) [reply group-generators/helpers/data-providers/snapshot/provider.ts:513]: Same here - this was explicit in the GitHub issue so I included it

Yes - this was explicit in the GitHub issue so I included it
https://github.com/sismo-core/sismo-hub/issues/1431 

```
{
"name": "number of proposals",
"argName": "TODO",
"type": "TODO",
"example": "3",
"description": "A minimum number of proposals"
},
```

File: group-generators/helpers/data-providers/snapshot/provider.ts [hunk 5]
```diff
@@ -216,4 +226,461 @@ export default class SnapshotProvider
...
+  /**
+   * Retrieves space followers from Snapshot.
+   * @param {Object} input - Input parameters.
+   * @param {string} input.space - The space to query followers for.
+   * @param {string} [input.date] - Optional date string to filter followers created after the specified date.
+   * @returns {Promise<FetchedData>} - A Promise that resolves to an object containing fetched data with follower addresses as keys and values set to 1.
+   */
+  public async querySpaceFollowers({
+    space,
+    date,
```
[the remark this answers is not in the document]
Comment (sampolgar) [reply group-generators/helpers/data-providers/snapshot/provider.ts:239]: Done

**Comment by sampolgar** [comment]
@MartinGbz regarding the count function in the dataprovider index file, does the interface `ISnapshotProvider` handle the method import/exports?

File: group-generators/helpers/data-providers/snapshot/provider.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (MartinGbz) [reply group-generators/helpers/data-providers/snapshot/provider.ts]: Maybe I didn't make myself clear sorry

First, no problem at all sir, it wasn't specified in the issue so it's not your fault ;))
It's just that I I very liked the fact that you use the "number of follows the user has" rather than a classic default value! So I instantly thought that it would be even better if the user can choose between this 2 options (default value or number of followers).

=> If the user doesn't set a defaultValue, by default the value of the account will be the number of vote.
So maybe the defaultValue is not the right name for this argument. Maybe "forcedValue" is better?

So it would be so nice to have this 2 choices for all the function that have a "special" account value (number of votes, number of proposal... etc).
In other words for these 4 functions:
```
querySpaceVoters
querySpaceAuthors
querySpaceVotersAboveX
queryProposalAuthorsAboveX
```

File: group-generators/helpers/data-providers/snapshot/provider.ts (hunk 3 above)
Comment (MartinGbz) [reply group-generators/helpers/data-providers/snapshot/provider.ts]: Yes so it's : "the minimum number of proposals created by an author" and no "the minimum number of votes for an author" like what you wrote

File: group-generators/helpers/data-providers/snapshot/provider.ts (hunk 2 above)
Comment (MartinGbz) [reply group-generators/helpers/data-providers/snapshot/provider.ts]: Yes so it's "number of votes for a voter" an not:  "number of votes for an author" like what you wrote isn't it?

Comment (MartinGbz) [reply group-generators/helpers/data-providers/snapshot/provider.ts]: Also I was thinking, wouldn't be better not to create another function for `querySpaceVotersAboveX` & `queryProposalAuthorsAboveX` but have directly another arguments in the `querySpaceVoters` and `querySpaceAuthors` functions that le the user choose if he want a special minimumValue. (but by default set it to 1).
The UX would be way nicer!
This issue has been created when the optional argument wasn't available in the Factory, but that that's not the case anymore, so it makes sens to do this. What do you think?

**Comment by MartinGbz** [comment]
> @MartinGbz regarding the count function in the dataprovider index file, does the interface `ISnapshotProvider` handle the method import/exports?

I don't really understand what you mean. The `ISnapshotProvider` is just a way to define the interface of the provider, but it's not required to do it.

File: group-generators/helpers/data-providers/snapshot/provider.ts (hunk 2 above)
Comment (sampolgar) [reply group-generators/helpers/data-providers/snapshot/provider.ts]: > Yes so it's "number of votes for a voter" an not: "number of votes for an author" like what you wrote isn't it?

Ok got it. I was confused what you were referring to. I'll update the comment

File: group-generators/helpers/data-providers/snapshot/provider.ts (hunk 3 above)
Comment (sampolgar) [reply group-generators/helpers/data-providers/snapshot/provider.ts]: Same here -  I was confused what you were referring to. I'll update the comment

**Comment by sampolgar** [comment]
> Also, you just forgot to add your count function to the Data Provider index file (group-generators/helpers/data-providers/index.ts)

The Index file is importing the `ISnapshotProvider` and the rest are `Types`. The export contains `SnapshotProvider`, `ISnapshotProvider` and `Types`

```
import {
  ISnapshotProvider,
  QueryAllVotersInput,
  QueryProposalVotersInput,
  QuerySpaceVotersInput,
  QuerySpaceAuthorsInput,
  QuerySpaceFollowersInput,
  QuerySpaceAdminsInput,
  QuerySpaceVotersAboveXInput,
  QueryProposalAuthorsAboveXInput,
} from "./types";

export {
  SnapshotProvider,
  ISnapshotProvider,
  QueryAllVotersInput as SnapshotQueryAllVotersInput,
  QuerySpaceVotersInput as SnapshotQuerySpaceVotersInput,
  QueryProposalVotersInput as SnapshotQueryProposalVotersInput,
  QuerySpaceAuthorsInput,
  QuerySpaceFollowersInput,
  QuerySpaceAdminsInput,
  QuerySpaceVotersAboveXInput,
  QueryProposalAuthorsAboveXInput,
};
```

Should I be adding the functions here as well?

File: group-generators/helpers/data-providers/snapshot/provider.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (sampolgar) [reply group-generators/helpers/data-providers/snapshot/provider.ts]: Sure, a few questions

1. what use case would someone set the forced value?
2. do you want this in the factory interface? if so, can you specify the interface schema attributes please?

**Comment by MartinGbz** [comment]
> > Also, you just forgot to add your count function to the Data Provider index file (group-generators/helpers/data-providers/index.ts)
> 
> The Index file is importing the `ISnapshotProvider` and the rest are `Types`. The export contains `SnapshotProvider`, `ISnapshotProvider` and `Types`
> 
> ```
> import {
>   ISnapshotProvider,
>   QueryAllVotersInput,
>   QueryProposalVotersInput,
>   QuerySpaceVotersInput,
>   QuerySpaceAuthorsInput,
>   QuerySpaceFollowersInput,
>   QuerySpaceAdminsInput,
>   QuerySpaceVotersAboveXInput,
>   QueryProposalAuthorsAboveXInput,
> } from "./types";
> 
> export {
>   SnapshotProvider,
>   ISnapshotProvider,
>   QueryAllVotersInput as SnapshotQueryAllVotersInput,
>   QuerySpaceVotersInput as SnapshotQuerySpaceVotersInput,
>   QueryProposalVotersInput as SnapshotQueryProposalVotersInput,
>   QuerySpaceAuthorsInput,
>   QuerySpaceFollowersInput,
>   QuerySpaceAdminsInput,
>   QuerySpaceVotersAboveXInput,
>   QueryProposalAuthorsAboveXInput,
> };
> ```
> 
> Should I be adding the count functions here as well?

Oh okay, so yes you can also add the count functions to the interface 👍

File: group-generators/helpers/data-providers/snapshot/provider.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (MartinGbz) [reply group-generators/helpers/data-providers/snapshot/provider.ts]: 1. Let's take an example:
I want to get all the voters of the aave space AND get all the onchain voters of AAVE.
But I want to give onchain voters an advantage over snapshot voters, so I'll set a forcedValue of 1 for snapshot voter, and a value of 2 for onchain voters.

2. Yes 👍 And sure, where will be for example the schema of the `querySpaceAuthors`
```
    {
      "name": "Get authors of space",
      "functionName": "querySpaceAuthors",
      "countFunctionName": "querySpaceAuthorsCount",
      "description": "Returns all authors of a specific space on Snapshot",
      "args": [
        {
          "name": "space",
          "argName": "space",
          "type": "string",
          "example": "sismo.eth",
          "description": "A specific space identifier"
        },
        {
          "name": "Forced value",
          "argName": "forcedValue",
          "type": "number",
          "example": "1",
          "description": "Force a specific value for all the accounts fetched"
        }
      ]
    },
```
